README.md
# Introduction of Informeasure
The information theory R package named `Informeasure` is to quantify nonlinear dependence between variables in biological regulatory network inferences. This package compiles most of the information measures currently available: mutual information (MI), conditional mutual information (CMI)[1], interaction information (II)[2], partial information decomposition (PID)[3] and part mutual information (PMI)[4], all of which end with .measure() in form. They are `MI.measure()` for MI, `CMI.measure()` for CMI, `II.measure()` for II, `PID.measure()` for PID and `PMI.measure()` for PMI. The first estimator is used to infer bivariate networks while the last four are dedicated to analysis of trivariate networks. I here consider estimating information measures from breast cancer expression profile data generated by The Cancer Genome Atlas (TCGA), with applications in various types of transcriptome regulatory network inferences.

# Prerequisites and Installation
```{r echo=FALSE, results='hide', message=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# The following initializes usage of Bioc devel
BiocManager::install(version='devel')

BiocManager::install("Informeasure")
```

# Main functions demonstration
Information measure is typically implemented by first discretizing continuous variables into a count table, evaluating probability from the counting, and/or then estimating entropy according to the (joint) probability matrix, finally calculating the information value that is the most representative for the association between variables. Two of the most common discretization methods are adopted in this package. One is a uniform width-based method (default) that divides the continuous data into `N` count bins with equal width. The other alternative is a uniform frequency-based approach that determines the continuous data into `N` count bins with equal count number. By default in both methods, the number of bins in these two methods is initialized into a round-off value based on the square root of the data size. In the process of probability estimation, three types of probability estimators referencing to the `entropy` package[5] that include the empirical estimator (default), the Dirichlet distribution estimator and the shrinkage estimator, while the Dirichlet distribution estimator also includes four different distribution with different prior values. These different probability estimators are showed in detail below. 

method = "ML": empirical estimator, also referred to maximum likelihood estimator, 

method = "Jeffreys": Dirichlet distribution estimator with prior a = 0.5, 

method = "Laplace": Dirichlet distribution estimator with prior a = 1, 

method = "SG": Dirichlet distribution estimator with prior a = 1/length(count table), 

method = "minimax": Dirichlet distribution estimator with prior a = sqrt(sum(count table))/length(count table), 

method = "shrink": shrinkage estimator.

## MI.measure(): mutual information
In the case of two variables, the representative method is mutual information, used to measure the mutual dependence between two joint variables. It can be used to identify dependence between proteins in protein-protein interaction network inference. Two types of data formats can be used as input to the algorithm. One is the simple data.frame data type, and the other is the SummarizedExperiment data type.

```{r}
# data.frame data type 
library(Informeasure)

load(system.file("extdata/tcga.brca.testdata.Rdata", package = "Informeasure"))

mRNAexpression   <- log2(mRNAexpression + 1)

x <- as.numeric(mRNAexpression[which(rownames(mRNAexpression) == "BRCA1"), ])
y <- as.numeric(mRNAexpression[which(rownames(mRNAexpression) == "BARD1"), ])

XY <- discretize2D(x,y)

MI.measure(XY)
```

```{r, echo = FALSE, results = 'hide', warning = FALSE}
suppressPackageStartupMessages(library(SummarizedExperiment))
```

```{r}
# SummarizedExperiment data type 
library(Informeasure)
library(SummarizedExperiment)

load(system.file("extdata/tcga.brca.testdata.Rdata", package = "Informeasure"))

mRNAexpression <- as.matrix(mRNAexpression)
se.mRNAexpression = SummarizedExperiment(assays = list(mRNAexpression = mRNAexpression))

assays(se.mRNAexpression)[["log2"]] <- log2(assays(se.mRNAexpression)[["mRNAexpression"]]+1)

x <- assays(se.mRNAexpression["BRCA1", ])$log2
y <- assays(se.mRNAexpression["BARD1", ])$log2

XY <- discretize2D(x,y)

MI.measure(XY)
```

## CMI.measure(): conditional mutual informaiton 
In the three-variable case, the most classic method is conditional mutual information. It is widely used to evaluate the expected mutual information between two random variables conditioned on the third one. Such characteristics of conditional mutual information are fully applicable to the lncRNA-associated ceRNA network inference.
```{r}
# data.frame data type
library(Informeasure)

load(system.file("extdata/tcga.brca.testdata.Rdata", package = "Informeasure"))

lncRNAexpression <- log2(lncRNAexpression + 1)
miRNAexpression  <- log2(miRNAexpression + 1)
mRNAexpression   <- log2(mRNAexpression + 1)

x <- as.numeric(miRNAexpression[which(rownames(miRNAexpression) == "hsa-miR-26a-5p"), ])
y <- as.numeric(mRNAexpression[which(rownames(mRNAexpression) == "PTEN"), ])
z <- as.numeric(lncRNAexpression[which(rownames(lncRNAexpression) == "PTENP1"), ])

XYZ <- discretize3D(x,y,z)

CMI.measure(XYZ)
```

```{r}
# SummarizedExperiment data type
library(Informeasure)
library(SummarizedExperiment)

load(system.file("extdata/tcga.brca.testdata.Rdata", package="Informeasure"))

lncRNAexpression <- as.matrix(lncRNAexpression)
se.lncRNAexpression = SummarizedExperiment(assays = list(lncRNAexpression = lncRNAexpression))

miRNAexpression <- as.matrix(miRNAexpression)
se.miRNAexpression = SummarizedExperiment(assays = list(miRNAexpression = miRNAexpression))

mRNAexpression <- as.matrix(mRNAexpression)
se.mRNAexpression = SummarizedExperiment(assays = list(mRNAexpression = mRNAexpression))

assays(se.lncRNAexpression)[["log2"]] <- log2(assays(se.lncRNAexpression)[["lncRNAexpression"]] + 1)

assays(se.miRNAexpression)[["log2"]] <- log2(assays(se.miRNAexpression)[["miRNAexpression"]] + 1)

assays(se.mRNAexpression)[["log2"]] <- log2(assays(se.mRNAexpression)[["mRNAexpression"]] + 1)


x <- assays(se.miRNAexpression["hsa-miR-26a-5p", ])$log2
y <- assays(se.mRNAexpression["PTEN", ])$log2
z <- assays(se.lncRNAexpression["PTENP1", ])$log2

XYZ <- discretize3D(x,y,z)

CMI.measure(XYZ)
```

## II.measure(): interaction information
Interaction information, also known as co-information, measures the amount information contained in a set of variables beyond any subset of those variables. The number of variables here is limited to three. It can be applied to explore the cooperative or competitive regulation mechanism of two miRNAs on the common target mRNA. 
```{r}
# data.frame data type
library(Informeasure)

load(system.file("extdata/tcga.brca.testdata.Rdata", package = "Informeasure"))

miRNAexpression  <- log2(miRNAexpression + 1)
mRNAexpression   <- log2(mRNAexpression + 1)

x <- as.numeric(miRNAexpression[which(rownames(miRNAexpression) == "hsa-miR-34a-5p"), ])
y <- as.numeric(mRNAexpression[which(rownames(mRNAexpression) == "MYC"), ])
z <- as.numeric(miRNAexpression[which(rownames(miRNAexpression) == "hsa-miR-34b-5p"), ])

XYZ <- discretize3D(x,y,z)

II.measure(XYZ)
```

```{r}
# SummarizedExperiment data type
library(Informeasure)
library(SummarizedExperiment)

load(system.file("extdata/tcga.brca.testdata.Rdata", package="Informeasure"))

miRNAexpression <- as.matrix(miRNAexpression)
se.miRNAexpression = SummarizedExperiment(assays = list(miRNAexpression = miRNAexpression))

mRNAexpression <- as.matrix(mRNAexpression)
se.mRNAexpression = SummarizedExperiment(assays = list(mRNAexpression = mRNAexpression))

assays(se.miRNAexpression)[["log2"]] <- log2(assays(se.miRNAexpression)[["miRNAexpression"]] + 1)

assays(se.mRNAexpression)[["log2"]] <- log2(assays(se.mRNAexpression)[["mRNAexpression"]] + 1)

x <- assays(se.miRNAexpression["hsa-miR-34a-5p", ])$log2
y <- assays(se.mRNAexpression["MYC", ])$log2
z <- assays(se.miRNAexpression["hsa-miR-34b-5p", ])$log2

XYZ <- discretize3D(x,y,z)

II.measure(XYZ)
```
## PID.measure(): partial information decomposition
Partial information decomposition decomposes two source information acting on the common target into four information parts: joint information (synergy), unique information from x, unique information from y and shared information (redundancy). It also can be applied to explore the cooperative or competitive regulation mechanism of two miRNAs on the common target mRNA.
```{r}
# data.frame data type
library(Informeasure)

load(system.file("extdata/tcga.brca.testdata.Rdata", package = "Informeasure"))

miRNAexpression  <- log2(miRNAexpression + 1)
mRNAexpression   <- log2(mRNAexpression + 1)

x <- as.numeric(miRNAexpression[which(rownames(miRNAexpression) == "hsa-miR-34a-5p"), ])
y <- as.numeric(miRNAexpression[which(rownames(miRNAexpression) == "hsa-miR-34b-5p"), ])
z <- as.numeric(mRNAexpression[which(rownames(mRNAexpression) == "MYC"), ])

XYZ <- discretize3D(x,y,z)

PID.measure(XYZ)
```

```{r}
# SummarizedExperiment data type
library(Informeasure)
library(SummarizedExperiment)

load(system.file("extdata/tcga.brca.testdata.Rdata", package="Informeasure"))

miRNAexpression <- as.matrix(miRNAexpression)
se.miRNAexpression = SummarizedExperiment(assays = list(miRNAexpression = miRNAexpression))

mRNAexpression <- as.matrix(mRNAexpression)
se.mRNAexpression = SummarizedExperiment(assays = list(mRNAexpression = mRNAexpression))

assays(se.miRNAexpression)[["log2"]] <- log2(assays(se.miRNAexpression)[["miRNAexpression"]] + 1)

assays(se.mRNAexpression)[["log2"]] <- log2(assays(se.mRNAexpression)[["mRNAexpression"]] + 1)

x <- assays(se.miRNAexpression["hsa-miR-34a-5p", ])$log2
y <- assays(se.miRNAexpression["hsa-miR-34b-5p", ])$log2
z <- assays(se.mRNAexpression["MYC", ])$log2

XYZ <- discretize3D(x,y,z)

PID.measure(XYZ)
```
## PMI.measure(): part mutual information 
Part mutual information devotes to measuring the non-linearly direct dependencies between two random variables given a third, especially when any one variable has a potentially strong correlation with the third one. Such characteristics of part mutual information are also fully applicable to the lncRNA-associated ceRNA network inference.
```{r}
# data.frame data type
library(Informeasure)

load(system.file("extdata/tcga.brca.testdata.Rdata", package = "Informeasure"))

lncRNAexpression <- log2(lncRNAexpression + 1)
miRNAexpression  <- log2(miRNAexpression + 1)
mRNAexpression   <- log2(mRNAexpression + 1)

x <- as.numeric(miRNAexpression[which(rownames(miRNAexpression)   == "hsa-miR-26a-5p"), ])
y <- as.numeric(mRNAexpression[which(rownames(mRNAexpression)     == "PTEN"), ])
z <- as.numeric(lncRNAexpression[which(rownames(lncRNAexpression) == "PTENP1"), ])

XYZ <- discretize3D(x,y,z)

PMI.measure(XYZ)
```

```{r}
# SummarizedExperiment data type 
library(Informeasure)
library(SummarizedExperiment)

load(system.file("extdata/tcga.brca.testdata.Rdata", package="Informeasure"))

lncRNAexpression <- as.matrix(lncRNAexpression)
se.lncRNAexpression = SummarizedExperiment(assays = list(lncRNAexpression = lncRNAexpression))

miRNAexpression <- as.matrix(miRNAexpression)
se.miRNAexpression = SummarizedExperiment(assays = list(miRNAexpression = miRNAexpression))

mRNAexpression <- as.matrix(mRNAexpression)
se.mRNAexpression = SummarizedExperiment(assays = list(mRNAexpression = mRNAexpression))

assays(se.lncRNAexpression)[["log2"]] <- log2(assays(se.lncRNAexpression)[["lncRNAexpression"]] + 1)

assays(se.miRNAexpression)[["log2"]] <- log2(assays(se.miRNAexpression)[["miRNAexpression"]] + 1)

assays(se.mRNAexpression)[["log2"]] <- log2(assays(se.mRNAexpression)[["mRNAexpression"]] + 1)

x <- assays(se.miRNAexpression["hsa-miR-26a-5p", ])$log2
y <- assays(se.mRNAexpression["PTEN", ])$log2
z <- assays(se.lncRNAexpression["PTENP1", ])$log2

XYZ <- discretize3D(x,y,z)

PMI.measure(XYZ)
```

# Conclusions
This package provides implementations of five currently popular information measures. The base installation of this package allows users to approach these information measures to infer bivariate even multivariate biological regulatory networks. But please be noted that the provided package is not only limited to bioinformatics applications. Optionally other research fields can also employ this package to generally evaluate information relations between variables.

# Acknowledgement
I would like to thank Ms. Song Jing for her careful proofreading of the manuscript, Mr. Xianghua Wang for his helpful discussions on the PMI algorithm, and Dr. Junpeng Zhang, Mr. Nitesh Turaga and Mr. Martin Morgan for their informative suggestions on writing the R package. I also would like to thank my family for their persistent support during my most difficult times in 2020!


# License
This project is licensed under the GPL license.

# References
[1] Wyner A D. A definition of conditional mutual information for arbitrary ensembles[J]. Information & Computation, 1978, 38(1): 51-59.

[2] Mcgill W J. Multivariate information transmission[J]. Psychometrika, 1954, 19(2): 97-116. 

[3] Williams P L, Beer R D. Nonnegative Decomposition of Multivariate Information[J]. arXiv: Information Theory, 2010.

[4] Zhao J, Zhou Y, Zhang X, et al. Part mutual information for quantifying direct associations in networks[J]. Proceedings of the National Academy of Sciences of the United States of America, 2016, 113(18): 5130-5135.

[5] Hausser J. and Strimmer K. Entropy inference and the James-Stein estimator, with application to nonlinear gene association networks[J]. The Journal of Machine Learning Research, 2009, 10, 1469-1484.


NEWS
Changes in version 1.1.1 (2020-10-30)
-------------------------
o make changes

Changes in version 0.99.8 (2020-10-06)
-------------------------
o make changes

Changes in version 0.99.7 (2020-10-06)
-------------------------
o make changes

Changes in version 0.99.6 (2020-09-21)
-------------------------
o make changes


Changes in version 0.99.5 (2020-09-20)
-------------------------
o Fix the the issue reported in the build report where used "1:lenght"" instead of "seq_len"
o Use "<-" for assignment rather than "="
o Fix indentation


Changes in version 0.99.4 (2020-08-14)
-------------------------
o Fix the warning of unnecessary package calls


Changes in version 0.99.3 (2020-08-14)
-------------------------
o Fix the warning of unnecessary package calls


Changes in version 0.99.2 (2020-08-14)
-------------------------
o Fix the warning of unnecessary package calls


Changes in version 0.99.1 (2020-08-13)
-------------------------
o Made the following significant changes
  o Added Informeasure.Rproj into the .gitignore file
  o Replaced Author/Maintainer with Authors@R in the DESCRIPTION file
  o Added a NEWS file
  o Added a .R file in tests/ directory
  o Update R version dependency from 3.5.0 to 4.0
  o Use TRUE/FALSE instead of T/F in PMI.plugin()


Changes in version 0.99.0 (2020-08-09)
-------------------------
o Submitted to Bioconductor


tests/testthat.R
library(testthat)

library(Informeasure)

test_check("Informeasure")


R/information.measure.R
#' A comprehensive function for evaluating mutual information
#'
#' The MI.measure function is used to calculate the mutual information between two random variables from the joint count table.
#'
#' Six probability estimation methods are available to evaluate the underlying bin probability from observed counts: \cr
#' method = "ML": maximum likelihood estimator, also referred to empirical probability, \cr
#' method = "Jeffreys": Dirichlet distribution estimator with prior a = 0.5, \cr
#' method = "Laplace": Dirichlet distribution estimator with prior a = 1, \cr
#' method = "SG": Dirichlet distribution estimator with prior a = 1/length(XY), \cr
#' method = "minimax": Dirichlet distribution estimator with prior a = sqrt(sum(XY))/length(XY), \cr
#' method = "shrink": shrinkage estimator.
#' @param XY a joint count distribution table of two random variables.
#' @param method six probability estimation algorithms are available, "ML" is the default.
#' @param lambda.probs the shrinkage intensity, only called when the probability estimator is "shrink".
#' @param unit the base of the logarithm. The default is natural logarithm, which is "log".
#' For evaluating entropy in bits, it is suggested to set the unit to "log2".
#' @param verbose a logic variable. if verbose is true, report the shrinkage intensity.
#'
#' @return MI.measure returns the mutual information.
#' @export
#' @import entropy
#' @importFrom entropy freqs.empirical freqs.Dirichlet freqs.shrink
#'
#' @references
#' Hausser, J., & Strimmer, K. (2009). Entropy Inference and the James-Stein Estimator, with Application to Nonlinear Gene Association Networks.
#' Journal of Machine Learning Research, 1469-1484.
#'
#' Wyner, A. D. (1978). A definition of conditional mutual information for arbitrary ensembles. Information & Computation, 38(1), 51-59.
#'
#' @examples
#' # two numeric vectors corresponding to two continuous random variables
#' x <- c(0.0, 0.2, 0.2, 0.7, 0.9, 0.9, 0.9, 0.9, 1.0)
#' y <- c(1.0, 2.0,  12, 8.0, 1.0, 9.0, 0.0, 3.0, 9.0)
#'
#' # corresponding joint count table estimated by "uniform width" algorithm
#' XY <- discretize2D(x, y, "uniform_width")
#'
#' # corresponding mutual information
#' MI.measure(XY)
MI.measure <- function(XY, method = c("ML", "Jeffreys", "Laplace", "SG", "minimax", "shrink"),
                       lambda.probs, unit = c("log", "log2", "log10"), verbose = TRUE){

  method <- match.arg(method)
    unit <- match.arg(unit)

  if(method == "ML"){
    probs <- freqs.empirical(XY)
    MI <- MI.plugin(probs, unit)
  }

  if(method == "Jeffreys"){
    probs <- freqs.Dirichlet(XY, a = 1/2)
    MI <- MI.plugin(probs, unit)
  }

  if(method == "Laplace"){
    probs <- freqs.Dirichlet(XY, a = 1)
    MI <- MI.plugin(probs, unit)
  }

  if(method == "SG"){
    probs <- freqs.Dirichlet(XY, a = 1/length(XY))
    MI <- MI.plugin(probs, unit)
  }

  if(method == "minimax"){
    probs <- freqs.Dirichlet(XY, a = sqrt(sum(XY))/length(XY))
    MI <- MI.plugin(probs, unit)
  }

  if(method == "shrink"){
    probs <- freqs.shrink(XY, lambda.freqs = lambda.probs, verbose = verbose)
    MI <- MI.plugin(probs, unit)
  }

  return(MI)
}


#' A comprehensive function for estimating conditional mutual information
#'
#' The CMI.measure function is used to calculate the expected mutual information between two random variables conditioned on the third one
#' from the joint count table.
#'
#' Six probability estimation methods are available to evaluate the underlying bin probability from observed counts: \cr
#' method = "ML": maximum likelihood estimator, also referred to empirical probability, \cr
#' method = "Jeffreys": Dirichlet distribution estimator with prior a = 0.5, \cr
#' method = "Laplace": Dirichlet distribution estimator with prior a = 1, \cr
#' method = "SG": Dirichlet distribution estimator with prior a = 1/length(XY), \cr
#' method = "minimax": Dirichlet distribution estimator with prior a = sqrt(sum(XY))/length(XY), \cr
#' method = "shrink": shrinkage estimator.
#' @param XYZ a joint count distribution table of three random variables.
#' @param method six probability estimation algorithms are available, "ML" is the default.
#' @param lambda.probs  the shrinkage intensity, only called when the probability estimator is "shrink".
#' @param unit the base of the logarithm. The default is natural logarithm, which is "log".
#' For evaluating entropy in bits, it is suggested to set the unit to "log2".
#' @param verbose a logic variable. if verbose is true, report the shrinkage intensity.
#'
#' @return CMI.measure returns the conditional mutual information.
#' @export
#' @import entropy
#' @importFrom entropy freqs.empirical freqs.Dirichlet freqs.shrink
#'
#' @references
#' #' Hausser, J., & Strimmer, K. (2009). Entropy Inference and the James-Stein Estimator, with Application to Nonlinear Gene Association Networks.
#' Journal of Machine Learning Research, 1469-1484.
#'
#' @examples
#' # three numeric vectors corresponding to three continuous random variables
#' x <- c(0.0, 0.2, 0.2, 0.7, 0.9, 0.9, 0.9, 0.9, 1.0)
#' y <- c(1.0, 2.0,  12, 8.0, 1.0, 9.0, 0.0, 3.0, 9.0)
#' z <- c(3.0, 7.0, 2.0,  11,  10,  10,  14, 2.0,  11)
#'
#' # corresponding joint count table estimated by "uniform width" algorithm
#' XYZ <- discretize3D(x, y, z, "uniform_width")
#'
#' # corresponding conditional mutual information
#' CMI.measure(XYZ)
CMI.measure <- function(XYZ, method = c("ML", "Jeffreys", "Laplace", "SG", "minimax", "shrink"),
                       lambda.probs, unit = c("log", "log2", "log10"), verbose = TRUE){

  method <- match.arg(method)
    unit <- match.arg(unit)

  if(method == "ML"){
    probs <- freqs.empirical(XYZ)
    CMI <- CMI.plugin(probs, unit)
  }

  if(method == "Jeffreys"){
    probs <- freqs.Dirichlet(XYZ, a = 1/2)
    CMI <- CMI.plugin(probs, unit)
  }

  if(method == "Laplace"){
    probs <- freqs.Dirichlet(XYZ, a = 1)
    CMI <- CMI.plugin(probs, unit)
  }

  if(method == "SG"){
    probs <- freqs.Dirichlet(XYZ, a = length(XYZ))
    CMI <- CMI.plugin(probs, unit)
  }

  if(method == "minimax"){
    probs <- freqs.Dirichlet(XYZ, a = sqrt(sum(XYZ))/length(XYZ))
    CMI <- CMI.plugin(probs, unit)
  }

  if(method == "shrink"){
    probs <- freqs.shrink(XYZ, lambda.freqs = lambda.probs, verbose = verbose)
    CMI <- CMI.plugin(probs, unit)
  }

  return(CMI)
}


#' A comprehensive function for evaluating interaction information
#'
#' The II.measure function is used to calculate the amount information contained in a set of variables from the joint count table.
#' The number of variables here is limited to three.
#'
#' Six probability estimation methods are available to evaluate the underlying bin probability from observed counts: \cr
#' method = "ML": maximum likelihood estimator, also referred to empirical probability, \cr
#' method = "Jeffreys": Dirichlet distribution estimator with prior a = 0.5, \cr
#' method = "Laplace": Dirichlet distribution estimator with prior a = 1, \cr
#' method = "SG": Dirichlet distribution estimator with prior a = 1/length(XY), \cr
#' method = "minimax": Dirichlet distribution estimator with prior a = sqrt(sum(XY))/length(XY), \cr
#' method = "shrink": shrinkage estimator.
#' @param XYZ a joint count distribution table of three random variables.
#' @param method six probability estimation algorithms are available, "ML" is the default.
#' @param lambda.probs  the shrinkage intensity, only called when the probability estimator is "shrink".
#' @param unit the base of the logarithm. The default is natural logarithm, which is "log".
#' For evaluating entropy in bits, it is suggested to set the unit to "log2".
#' @param verbose a logic variable. if verbose is true, report the shrinkage intensity.
#'
#' @return II.measure returns the interaction information.
#' @export
#' @import entropy
#' @importFrom entropy freqs.empirical freqs.Dirichlet freqs.shrink
#'
#' @references
#' Hausser, J., & Strimmer, K. (2009). Entropy Inference and the James-Stein Estimator, with Application to Nonlinear Gene Association Networks.
#' Journal of Machine Learning Research, 1469-1484.
#'
#' Mcgill, W. J. (1954). Multivariate information transmission. Psychometrika, 19(2), 97-116.
#'
#' @examples
#' # three numeric vectors corresponding to three continuous random variables
#' x <- c(0.0, 0.2, 0.2, 0.7, 0.9, 0.9, 0.9, 0.9, 1.0)
#' y <- c(1.0, 2.0,  12, 8.0, 1.0, 9.0, 0.0, 3.0, 9.0)
#' z <- c(3.0, 7.0, 2.0,  11,  10,  10,  14, 2.0,  11)
#'
#' # corresponding joint count table estimated by "uniform width" algorithm
#' XYZ <- discretize3D(x, y, z, "uniform_width")
#'
#' # corresponding interaction information
#' II.measure(XYZ)
II.measure <- function(XYZ, method = c("ML", "Jeffreys", "Laplace", "SG", "minimax", "shrink"),
                      lambda.probs, unit = c("log", "log2", "log10"), verbose = TRUE){

  method <- match.arg(method)
    unit <- match.arg(unit)

  if(method == "ML"){
    probs <- freqs.empirical(XYZ)
    II <- II.plugin(probs, unit)
  }

  if(method == "Jeffreys"){
    probs <- freqs.Dirichlet(XYZ, a = 1/2)
    II <- II.plugin(probs, unit)
  }

  if(method == "Laplace"){
    probs <- freqs.Dirichlet(XYZ, a = 1)
    II <- II.plugin(probs, unit)
  }

  if(method == "SG"){
    probs <- freqs.Dirichlet(XYZ, a = length(XYZ))
    II <- II.plugin(probs, unit)
  }

  if(method == "minimax"){
    probs <- freqs.Dirichlet(XYZ, a = sqrt(sum(XYZ))/length(XYZ))
    II <- II.plugin(probs, unit)
  }

  if(method == "shrink"){
    probs <- freqs.shrink(XYZ, lambda.freqs = lambda.probs, verbose = verbose)
    II <- II.plugin(probs, unit)
  }

  return(II)
}


#' A comprehensive function for evaluating the partial information decomposition
#'
#' The PID.measure function is used to decompose two source information acting on the common target into four parts: joint information (synergy),
#' unique information from source x, unique information from source y and shared information (redundancy). The input of the PID.measure is the
#' joint count table.
#'
#' Six probability estimation methods are available to evaluate the underlying bin probability from observed counts: \cr
#' method = "ML": maximum likelihood estimator, also referred to empirical probability, \cr
#' method = "Jeffreys": Dirichlet distribution estimator with prior a = 0.5, \cr
#' method = "Laplace": Dirichlet distribution estimator with prior a = 1, \cr
#' method = "SG": Dirichlet distribution estimator with prior a = 1/length(XY), \cr
#' method = "minimax": Dirichlet distribution estimator with prior a = sqrt(sum(XY))/length(XY), \cr
#' method = "shrink": shrinkage estimator.
#' @param XYZ a joint count distribution table of three random variables
#' @param method six probability estimation algorithms are available, "ML" is the default.
#' @param lambda.probs  the shrinkage intensity, only called when the probability estimator is "shrink".
#' @param unit the base of the logarithm. The default is natural logarithm, which is "log".
#' For evaluating entropy in bits, it is suggested to set the unit to "log2".
#' @param verbose a logic variable. if verbose is true, report the shrinkage intensity.
#'
#' @return PID.measure returns a list that includes synergistic information, unique information from x, unique information from y,
#' redundant information and the sum of the four parts of information.
#' @export
#' @import entropy
#' @importFrom entropy freqs.empirical freqs.Dirichlet freqs.shrink
#'
#' @references
#' Hausser, J., & Strimmer, K. (2009). Entropy Inference and the James-Stein Estimator, with Application to Nonlinear Gene Association Networks.
#' Journal of Machine Learning Research, 1469-1484.
#'
#' Williams, P. L., & Beer, R. D. (2010). Nonnegative Decomposition of Multivariate Information. arXiv: Information Theory.
#'
#' Chan, T. E., Stumpf, M. P., & Babtie, A. C. (2017). Gene Regulatory Network Inference from Single-Cell Data Using Multivariate Information Measures.
#' Cell Systems, 5(3).
#'
#' @examples
#' # three numeric vectors corresponding to three continuous random variables
#' x <- c(0.0, 0.2, 0.2, 0.7, 0.9, 0.9, 0.9, 0.9, 1.0)
#' y <- c(1.0, 2.0,  12, 8.0, 1.0, 9.0, 0.0, 3.0, 9.0)
#' z <- c(3.0, 7.0, 2.0,  11,  10,  10,  14, 2.0,  11)
#'
#' # corresponding joint count table estimated by "uniform width" algorithm
#' XYZ <- discretize3D(x, y, z, "uniform_width")
#'
#' # corresponding partial information decomposition
#' PID.measure(XYZ)
#'
#' # corresponding count table estimated by "uniform frequency" algorithm
#' XYZ <- discretize3D(x, y, z, "uniform_frequency")
#'
#' # corresponding partial information decomposition
#' PID.measure(XYZ)
PID.measure <- function(XYZ, method = c("ML", "Jeffreys", "Laplace", "SG", "minimax",  "shrink"),
                       lambda.probs, unit = c("log", "log2", "log10"), verbose = TRUE){

  method <- match.arg(method)
    unit <- match.arg(unit)

  if(method == "ML"){
    probs <- freqs.empirical(XYZ)
    PID <- PID.plugin(probs, unit)
  }

  if(method == "Jeffreys"){
    probs <- freqs.Dirichlet(XYZ, a = 1/2)
    PID <- PID.plugin(probs, unit)
  }

  if(method == "Laplace"){
    probs <- freqs.Dirichlet(XYZ, a = 1)
    PID <- PID.plugin(probs, unit)
  }

  if(method == "SG"){
    probs <- freqs.Dirichlet(XYZ, a = length(XYZ))
    PID <- PID.plugin(probs, unit)
  }

  if(method == "minimax"){
    probs <- freqs.Dirichlet(XYZ, sqrt(sum(XYZ))/length(XYZ))
    PID <- PID.plugin(probs, unit)
  }

  if(method == "shrink"){
    probs <- freqs.shrink(XYZ, lambda.freqs = lambda.probs, verbose = verbose)
    PID <- PID.plugin(probs, unit)
  }

  return(PID)
}


#' A comprehensive function for evaluating part mutual information
#'
#'
#' The PMI.measure function is used to calculate the non-linearly direct dependencies between two variables conditioned on the third one
#' form the joint count table.
#'
#' Six probability estimation methods are available to evaluate the underlying bin probability from observed counts: \cr
#' method = "ML": maximum likelihood estimator, also referred to empirical probability, \cr
#' method = "Jeffreys": Dirichlet distribution estimator with prior a = 0.5, \cr
#' method = "Laplace": Dirichlet distribution estimator with prior a = 1, \cr
#' method = "SG": Dirichlet distribution estimator with prior a = 1/length(XY), \cr
#' method = "minimax": Dirichlet distribution estimator with prior a = sqrt(sum(XY))/length(XY), \cr
#' method = "shrink": shrinkage estimator.
#' @param XYZ a joint count distribution table of three random variables.
#' @param method six probability estimation algorithms are available, "ML" is the default.
#' @param lambda.probs  the shrinkage intensity, only called when the probability estimator is "shrink".
#' @param unit the base of the logarithm. The default is natural logarithm, which is "log".
#' For evaluating entropy in bits, it is suggested to set the unit to "log2".
#' @param verbose a logic variable. if verbose is true, report the shrinkage intensity.
#'
#' @return PMI.measure returns the part mutual information.
#' @export
#' @import entropy
#' @importFrom entropy freqs.empirical freqs.Dirichlet freqs.shrink
#'
#' @references
#' Hausser, J., & Strimmer, K. (2009). Entropy Inference and the James-Stein Estimator, with Application to Nonlinear Gene Association Networks.
#' Journal of Machine Learning Research, 1469-1484.
#'
#' Zhao, J., Zhou, Y., Zhang, X., & Chen, L. (2016). Part mutual information for quantifying direct associations in networks.
#' Proceedings of the National Academy of Sciences of the United States of America, 113(18), 5130-5135.
#'
#' @examples
#' # three numeric vectors corresponding to three continuous random variables
#' x <- c(0.0, 0.2, 0.2, 0.7, 0.9, 0.9, 0.9, 0.9, 1.0)
#' y <- c(1.0, 2.0,  12, 8.0, 1.0, 9.0, 0.0, 3.0, 9.0)
#' z <- c(3.0, 7.0, 2.0,  11,  10,  10,  14, 2.0,  11)
#'
#' # corresponding joint count table estimated by "uniform width" algorithm
#' XYZ <- discretize3D(x, y, z, "uniform_width")
#'
#' # corresponding part mutual information
#' PMI.measure(XYZ)
PMI.measure <- function(XYZ, method = c("ML", "Jeffreys", "Laplace", "SG", "minimax",  "shrink"),
                       lambda.probs, unit = c("log", "log2", "log10"), verbose = TRUE){

  method <- match.arg(method)
    unit <- match.arg(unit)

  if(method == "ML"){
    probs <- freqs.empirical(XYZ)
    PMI <- PMI.plugin(probs, unit)
  }

  if(method == "Jeffreys"){
    probs <- freqs.Dirichlet(XYZ, a = 1/2)
    PMI <- PMI.plugin(probs, unit)
  }

  if(method == "Laplace"){
    probs <- freqs.Dirichlet(XYZ, a = 1)
    PMI <- PMI.plugin(probs, unit)
  }

  if(method == "SG"){
    probs <- freqs.Dirichlet(XYZ, a = length(XYZ))
    PMI <- PMI.plugin(probs, unit)
  }

  if(method == "minimax"){
    probs <- freqs.Dirichlet(XYZ, sqrt(sum(XYZ))/length(XYZ))
    PMI <- PMI.plugin(probs, unit)
  }

  if(method == "shrink"){
    probs <- freqs.shrink(XYZ, lambda.freqs = lambda.probs, verbose = verbose)
    PMI <- PMI.plugin(probs, unit)
  }

  return(PMI)
}
